bin/kmstool-enclave-cli/main.c

```diff
@@ -0,0 +1,218 @@
+#include <aws/nitro_enclaves/kms.h>
+#include <aws/nitro_enclaves/nitro_enclaves.h>
+
+#include <aws/common/command_line_parser.h>
+#include <aws/common/encoding.h>
+#include <aws/common/logging.h>
+
+#include <json-c/json.h>
+
+#include <linux/vm_sockets.h>
+#include <sys/socket.h>
+
+#include <errno.h>
+#include <unistd.h>
+
+#define PROXY_PORT 8000
+
+enum status {
+    STATUS_OK,
+    STATUS_ERR,
+};
+
+#define fail_on(cond)                                                                                                  \
+    if (cond) {                                                                                                        \
+        return;                                                                                                        \
+    }
+
+struct app_ctx {
+    /* Allocator to use for memory allocations. */
+    struct aws_allocator *allocator;
+    /* KMS region to use. */
+    const struct aws_string *region;
+    /* vsock port on which to open service. */
+    uint32_t port;
+    /* vsock port on which vsock-proxy is available in parent. */
+    uint32_t proxy_port;
+
+    const struct aws_string *aws_access_key_id;
+    const struct aws_string *aws_secret_access_key;
+    const struct aws_string *aws_session_token;
+
+    const struct aws_string *ciphertext_b64;
+};
+
+static void s_usage(int exit_code) {
+    fprintf(stderr, "usage: enclave_server [options]\n");
+    fprintf(stderr, "\n Options: \n\n");
+    fprintf(stderr, "    --region REGION: AWS region to use for KMS\n");
+    fprintf(stderr, "    --proxy-port PORT: Connect to KMS proxy on PORT. Default: 8000\n");
+    fprintf(stderr, "    --aws-access-key-id ACCESS_KEY_ID: AWS access key ID\n");
+    fprintf(stderr, "    --aws-secret-access-key SECRET_ACCESS_KEY: AWS secret access key\n");
+    fprintf(stderr, "    --aws-session-token SESSION_TOKEN: Session token associated with the access key ID\n");
+    fprintf(stderr, "    --ciphertext CIPHERTEXT: base64-encoded ciphertext that need to decrypt\n");
+    fprintf(stderr, "    --help: Display this message and exit");
+    exit(exit_code);
+}
+
+static struct aws_cli_option s_long_options[] = {
+    {"region", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'r'},
+    {"proxy-port", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'x'},
+    {"help", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'h'},
+    {"aws-access-key-id", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'k'},
+    {"aws-secret-access-key", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 's'},
+    {"aws-session-token", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 't'},
+    {"ciphertext", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'c'},
+    {NULL, 0, NULL, 0},
+};
+
+static void s_parse_options(int argc, char **argv, struct app_ctx *ctx) {
+    ctx->proxy_port = PROXY_PORT;
+    ctx->region = NULL;
+
```

Comment:
Here you need to set `ctx->aws_access_key_id`, `ctx->aws_secret_access_key` and `ctx->aws_session_token` to NULL or otherwise set the whole structure to zero via `AWS_ZERO_STRUCT`.

Reply:
Initialize all `aws_string` variables